### ISSUE TITLE
Can anyone help why the loss backward does not work?

### ISSUE THREAD

**Issue by jbgao** [cvxgrp/cvxpylayers#112]
It seems that the program restarts the kernel when it hits test_loss.backward().   Anyone can give a hand on the issue?
J.

import cvxpy as cp
import matplotlib.pyplot as plt
import numpy as np
from scipy.linalg import orth
import torch
from cvxpylayers.torch import CvxpyLayer
torch.set_default_dtype(torch.double)

 
from sklearn.preprocessing import normalize

def gen_union_of_subspaces(ambient_dim, subspace_dim, num_subspaces, num_points_per_subspace, noise_level=0.0):
    """This funtion generates a union of subspaces under random model, i.e., 
    subspaces are independently and uniformly distributed in the ambient space,
    data points are independently and uniformly distributed on the unit sphere of each subspace
    Parameters
    -----------
    ambient_dim : int
        Dimention of the ambient space
    subspace_dim : int
        Dimension of each subspace (all subspaces have the same dimension)
    num_subspaces : int
        Number of subspaces to be generated
    num_points_per_subspace : int
        Number of data points from each of the subspaces
    noise_level : float
        Amount of Gaussian noise on data
		
    Returns
    -------
    data : shape (num_subspaces * num_points_per_subspace) by ambient_dim
        Data matrix containing points drawn from a union of subspaces as its rows
    label : shape (num_subspaces * num_points_per_subspace)
        Membership of each data point to the subspace it lies in
    """

    data = np.empty((num_points_per_subspace* num_subspaces, ambient_dim))
    label = np.empty(num_points_per_subspace * num_subspaces, dtype=int)
  
    for i in range(num_subspaces):
        basis = np.random.normal(size=(ambient_dim, subspace_dim))
        basis = orth(basis)
        coeff = np.random.normal(size=(subspace_dim, num_points_per_subspace))
        coeff = normalize(coeff, norm='l2', axis=0, copy=False)
        data_per_subspace = np.matmul(basis, coeff).T

        base_index = i*num_points_per_subspace
        data[(0+base_index):(num_points_per_subspace+base_index), :] = data_per_subspace
        label[0+base_index:num_points_per_subspace+base_index,] = i

    data += np.random.normal(size=(num_points_per_subspace * num_subspaces, ambient_dim)) * noise_level
  
    return data, label

# =================================================
# Generate dataset where data is drawn from a union of subspaces
# =================================================
ambient_dim = 9
subspace_dim = 6
num_subspaces = 5
num_points_per_subspace = 50

data, label = gen_union_of_subspaces(ambient_dim, subspace_dim, num_subspaces, num_points_per_subspace, 0.00)

m = data.shape[0] 
X = cp.Parameter((data.shape[0], ambient_dim))
Z = cp.Variable((data.shape[0],data.shape[0]))
lam = cp.Parameter(nonneg=True)
 

# set up objective
loss = cp.sum(cp.square(Z @ X - X))
reg = lam * cp.norm1(Z)  
objective = loss + reg
prob = cp.Problem(cp.Minimize(objective), [cp.diag(Z) == 0])
#prob.solve()
fit_lr = CvxpyLayer(prob, [X, lam], [Z])

#fit_lr(torch.from_numpy(data), torch.tensor(1.0))   # This works

lam_vals = np.logspace(-3, 2, 200)
test_losses = []
grads = []
for lam_val in lam_vals:
    lam_tch = torch.tensor([lam_val], requires_grad=True)
    lam_tch.grad = None
    Z_tch = fit_lr(torch.from_numpy(data), lam_tch)
    test_loss = (Z_tch[0] @ torch.from_numpy(data) - torch.from_numpy(data)).pow(2).mean()
    test_loss.backward()
    test_losses.append(test_loss.item())
    grads.append(lam_tch.grad.item())
    print(lam_val)

plt.semilogx()
plt.plot(lam_vals, test_losses, label='test loss')
plt.plot(lam_vals, grads, label='analytical gradient')
plt.plot(lam_vals[:-1], np.diff(test_losses) / np.diff(lam_vals), label='numerical gradient', linestyle='--')
plt.legend()
plt.xlabel("$\\lambda$")
plt.show()
